GithubRepo/Scenes/Search/View/SearchViewController.swift

```diff
@@ -85,74 +88,72 @@ class SearchViewController: UIViewController {
     // MARK: - Helper Methods
 
     private func handleSearchInitialState() {
+        self.spinner.stopAnimating()
         self.stateView?.isHidden = false
         self.stateTextView?.text = kInitialSearchStateText
         self.stateImageView?.image = UIImage(named: "Bookmark")
     }
 
     private func handleSearchEmptyState() {
+        self.spinner.stopAnimating()
         self.stateView?.isHidden = false
         self.stateTextView?.text = kEmptySearchStateText
         self.stateImageView?.image = UIImage(named: "BookmarkMad")
     }
 
     private func handleSearchErrorState() {
+        self.spinner.stopAnimating()
         self.stateView?.isHidden = false
         self.stateTextView?.text = kErrorSearchStateText
         self.stateImageView?.image = UIImage(named: "BookmarkError")
     }
-}
-
-// MARK: - Notifications from View Model
 
-extension SearchViewController: SearchViewModelDelegate {
-    func onChangeSearchError(error: Bool) {
-        DispatchQueue.main.async {
-            if error {
-                self.handleSearchErrorState()
-            }
-        }
-    }
-
-    func onChangeSearchLoadingState(isLoading: Bool) {
-        DispatchQueue.main.async {
-            if isLoading {
-                self.spinner.startAnimating()
-            } else {
-                self.spinner.stopAnimating()
-            }
-        }
+    private func handleSearchLoadingState() {
+        self.spinner.startAnimating()
+        self.stateView?.isHidden = true
     }
 
-    func onChangeSearchRepository(repoCellViewModels: [RepositoryCellViewModel]) {
-        DispatchQueue.main.async {
-            if repoCellViewModels.isEmpty {
-                self.handleSearchEmptyState()
-            } else {
-                self.stateView?.isHidden = true
-                self.tableView?.reloadData()
-            }
-        }
+    private func handleSearchContentState() {
+        self.spinner.stopAnimating()
+        self.stateView?.isHidden = true
     }
 }
 
-// MARK: - Table View Extension
-
-extension SearchViewController: UITableViewDataSource, UITableViewDelegate {
-    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return searchViewModel?.repositoryCellViewModels.count ?? 0
+// MARK: - Handle Notifications from View Model
+
+extension SearchViewController {
+    func subscribeSearchState() {
+        searchViewModel?.state.subscribe(onNext: { value -> Void in
+            DispatchQueue.main.async { [weak self] in
```

Comment:
this is a good example of using observable `Traits`. As you want to update the UI you must run in the `main` thread. Instead of calling `DispatchQueue.main.async` you can use the `Driver` observable. Basically, `Driver` is a type of `Observable` that:

- Can't error out.
- Observe occurs on main scheduler.
- Shares side effects (`share(replay: 1, scope: .whileConnected)`).

And you can easily transform your `Relay` into an `Driver` by calling `asDriver()`.
So it fits perfect here and basically for the most UI binds. So your code will look more like this: 
```suggestion
        searchViewModel?.state.asDriver()
            .drive { [weak self] in
```